### PR TITLE
fix(channel): handle writing to file instead of pipe

### LINKED ISSUE #19374: Segfault when `--headless rpcrequest(stioopen({}),...)` is redirected to file
### Neovim version (nvim -v)

NVIM v0.8.0-dev-nightly-12-g977790746

### Vim (not Nvim) behaves the same?

N/A

### Operating system/version

Linux r4 5.18.0-1-amd64 1 SMP PREEMPT_DYNAMIC Debian 5.18.2-1 (2022-06-06) x86_64 GNU/Linux

### Terminal name/version

N/A

### $TERM environment variable

N/A

### Installation

Appimage

### How to reproduce the issue

```
nvim --headless --cmd 'call rpcrequest(stdioopen({"rpc":v:true}),"nvim_command","echo hello")' >/tmp/tmp
```

### Expected behavior

Rpc request is written to /tmp/tmp.

### Actual behavior

Segfault without anything being written to the file. Removing the redirection makes the crash disappear.

Piping to `tee /tmp/tmp` instead of directly redirecting to a file is a suitable workaround for my usecase.

### LINKED ISSUE #30509: crash on chansend() write to redirected stdioopen() channel
### Problem

Segmentation fault when redirecting stdout to a file in `--headless`-mode.

`coredumpctl debug nvim` returns the following backtrace (when using `bt` in the gdb session that is being opened):

```
#0  uv__check_before_write (nbufs=1, stream=0x0, send_handle=0x0) at nvim/.deps/build/src/libuv/src/unix/stream.c:1304
#1  uv_write2 (req=0x5630ac7fe240, stream=0x0, bufs=0x7ffdef52ee60, nbufs=1, send_handle=0x0, cb=0x563087d29867 <write_cb>)
    at nvim/.deps/build/src/libuv/src/unix/stream.c:1342
#2  0x0000563087d26823 in wstream_write (stream=0x5630ac8143e8, buffer=0x5630ac813b20) at nvim/src/nvim/event/wstream.c:91
#3  0x0000563087cb7a2f in channel_send (id=<optimized out>, data=<optimized out>, len=3, data_owned=<optimized out>, error=0x7ffdef52eef8)
    at nvim/src/nvim/channel.c:621
#4  0x0000563087d04ae0 in f_chansend (argvars=0x7ffdef52f1e0, rettv=0x7ffdef52f180, fptr=...) at nvim/src/nvim/eval/funcs.c:680
#5  0x0000563087cfd65b in call_internal_func (fname=<optimized out>, argcount=<optimized out>, argvars=0x7ffdef52f1e0, rettv=0x7ffdef52f180)
    at nvim/src/nvim/eval/funcs.c:285
#6  0x0000563087d22cdd in call_func (funcname=0x7da316380420 "chansend", len=8, rettv=0x7ffdef52f180, argcount_in=<optimized out>, argvars_in=<optimized out>, funcexe=0x7ffdef52f1a0)
    at nvim/src/nvim/eval/userfunc.c:1745
#7  0x0000563087da6bc3 in nlua_call (lstate=0x7da31689e380) at nvim/src/nvim/lua/executor.c:1208
#8  0x0000563087f6e826 in lj_BC_FUNCC ()
#9  0x0000563087f5a0dd in lua_pcall (L=L@entry=0x7da31689e380, nargs=nargs@entry=0, nresults=nresults@entry=0, errfunc=errfunc@entry=-2) at lj_api.c:1122
#10 0x0000563087d9a2f9 in nlua_pcall (lstate=0x7da31689e380, nargs=0, nresults=0) at nvim/src/nvim/lua/executor.c:174
#11 0x0000563087da7f02 in nlua_exec_file (path=<optimized out>) at nvim/src/nvim/lua/executor.c:1876
#12 0x0000563087e60deb in do_source (fname=0x7ffdef530227 "minimal.lua", check_other=<optimized out>, is_vimrc=<optimized out>, ret_sid=<optimized out>)
    at nvim/src/nvim/runtime.c:2224
#13 0x0000563087db33ea in source_startup_scripts (parmp=0x7ffdef52f650) at nvim/src/nvim/main.c:2095
#14 0x0000563087dad272 in main (argc=5, argv=<optimized out>) at nvim/src/nvim/main.c:463

```

### Steps to reproduce

minimal.lua:
```lua
local channel = vim.fn.stdioopen { on_stdin = function() end }
vim.fn.chansend(channel, "123")
```

Running nvim:
```bash
nvim --headless --clean -u minimal.lua > stdout.txt
```

### Expected behavior

I would expect nvim not to segfault because of that.

### Nvim version (nvim -v)

v0.11.0-dev-828+g052875b9d

### Vim (not Nvim) behaves the same?

Vim doesn't use libuv as far as I know

### Operating system/version

Linux arch-book 6.10.7-arch1-1

### Terminal name/version

alacritty 0.13.2 (bb8ea18e)/tmux 3.4

### $TERM environment variable

tmux-256color

### Installation

build from repo

### PR DESCRIPTION
**Pull request by zeertzjq**
Fix #19374
Fix #30509

### CODE REVIEW

File: src/nvim/event/wstream.c [hunk 1]
```diff
@@ -73,6 +73,26 @@ bool wstream_write(Stream *stream, WBuffer *buffer)
...
+    uv_fs_t req;
+
+    // Synchronous write
+    uv_fs_write(stream->uv.idle.loop, &req, stream->fd, &uvbuf, 1, stream->fpos, NULL);
+
+    uv_fs_req_cleanup(&req);
+
+    wstream_release_wbuffer(buffer);
+
+    assert(stream->write_cb == NULL);
```
Comment (justinmk) [inline src/nvim/event/wstream.c:90]: should this go just after the if condition (line 80)?

Comment (zeertzjq) [reply src/nvim/event/wstream.c:90]: I was trying to match the order in the `write_cb()` below, where `write_cb` is checked after calling `wstream_release_buffer()`, and `closed` is checked afterwards. When `write_cb` is `NULL`, `closed` won't be set here, so `wstream_release_wbuffer()` is the only one needed here. An alternative is to move that part of `write_cb()` into a separate function, but then the `pending_reqs--` needs to be paired with a `pending_reqs++`, which I'm not sure where to put in this branch.

File: src/nvim/event/wstream.c [hunk 2]
```diff
@@ -73,6 +73,26 @@ bool wstream_write(Stream *stream, WBuffer *buffer)
   // This should not be called after a stream was freed
   assert(!stream->closed);
 
+  uv_buf_t uvbuf;
+  uvbuf.base = buffer->data;
+  uvbuf.len = UV_BUF_LEN(buffer->size);
+
+  if (!stream->uvstream) {
```
Comment (justinmk) [inline src/nvim/event/wstream.c:80]: i'm trying to understand, do we indicate anywhere that the lack of `stream.uvstream` implies a `stream.fd` ?  Should this be documented on `struct stream` , or a brief comment here? Normally this decided by a discriminator field...

In other places, the lack of `stream.uvstream` is taken to imply the presence of `stream.uv.idle`. 

https://github.com/neovim/neovim/blob/f3b7444e6638095305d86fa96fe0b7407c9b1648/src/nvim/event/rstream.c#L47-L51

Comment (zeertzjq) [reply src/nvim/event/wstream.c:80]: If you look at `fread_idle_cb`, you'll see that it reads from `stream->s.fd`.

Comment (justinmk) [reply src/nvim/event/wstream.c:80]: good point, but this is all very subtle, no? A brief docstring would really help (and an assert couldn't hurt). 

I see your latest push, thanks!

File: src/nvim/event/stream.c [hunk 3]
```diff
@@ -44,6 +44,7 @@ int stream_set_blocking(int fd, bool blocking)
 void stream_init(Loop *loop, Stream *stream, int fd, uv_stream_t *uvstream)
   FUNC_ATTR_NONNULL_ARG(2)
 {
+  assert(uvstream == NULL ? fd >= 0 : fd < 0);
```
Comment (justinmk) [inline src/nvim/event/stream.c:47]:
```suggestion
  // The underlying stream is either a file or an arbitrary uv stream.
  assert(uvstream == NULL ? fd >= 0 : fd < 0);
```

**Review by justinmk — APPROVED** [review]
Wonderful!

**Pull request merged**